internal/controller/podschedulingpolicy_controller.go

```diff
@@ -0,0 +1,211 @@
+)
+
+// PodSchedulingPolicyReconciler reconciles a PodSchedulingPolicy object.
+type PodSchedulingPolicyReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=everest.percona.com,resources=podschedulingpolicies,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=everest.percona.com,resources=podschedulingpolicies/status,verbs=get;update;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// Modify the Reconcile function to compare the state specified by
+// the PodSchedulingPolicy object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.1/pkg/reconcile
+func (r *PodSchedulingPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (rr ctrl.Result, rerr error) {
+	logger := log.FromContext(ctx)
+	logger.Info("Reconciling", "request", req)
+	defer func() {
+		logger.Info("Reconciled", "request", req)
+	}()
+
+	psp := &everestv1alpha1.PodSchedulingPolicy{}
+	if err := r.Get(ctx, req.NamespacedName, psp); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	dbList, err := common.DatabaseClustersThatReferenceObject(ctx, r.Client, podSchedulingPolicyNameField, "", psp.GetName())
+	if err != nil {
+		logger.Error(err, fmt.Sprintf("Failed to fetch DB clusters that use pod scheduling policy='%s'", psp.GetName()))
+		return ctrl.Result{}, errors.Join(err, fmt.Errorf("failed to fetch DB clusters that use pod scheduling policy='%s': %w", psp.GetName(), err))
+	}
+
+	// Update the status and finalizers of the PodSchedulingPolicy object after the reconciliation.
+	defer func() {
+		// Nothing to process on delete events
+		if !psp.GetDeletionTimestamp().IsZero() {
+			return
+		}
+
+		psp.Status.Used = len(dbList.Items) > 0
+		psp.Status.ObservedGeneration = psp.GetGeneration()
+		if err = r.Client.Status().Update(ctx, psp); err != nil {
+			rr = ctrl.Result{}
+			logger.Error(err, fmt.Sprintf("failed to update status for pod scheduling policy='%s'", psp.GetName()))
+			rerr = errors.Join(err, fmt.Errorf("failed to update status for pod scheduling policy='%s': %w", psp.GetName(), err))
+		}
+	}()
+
+	if err = r.ensureFinalizers(ctx, len(dbList.Items) > 0, psp); err != nil {
+		logger.Error(err, fmt.Sprintf("failed to update finalizers for pod scheduling policy='%s'", psp.GetName()))
+		return ctrl.Result{}, errors.Join(err, fmt.Errorf("failed to update finalizers for pod scheduling policy='%s': %w", psp.GetName(), err))
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *PodSchedulingPolicyReconciler) ensureFinalizers(ctx context.Context, used bool, psp *everestv1alpha1.PodSchedulingPolicy) error {
```

Comment:
Please add a comment describing how the finalizers are eventually removed for uninstallation of Everest